### ISSUE TITLE
ux_mode: "redirect" not working

### ISSUE THREAD

**Issue by SanFranciscoSunrise** [rubenCodeforges/ng-gapi#77]
When I include the ux_mode: "redirect" and redirect_uri: "http://localhost:4200/home" my app doesn't seem to redirect back to the correct path.

I know that after the google sign in screen there is some kind of redirect to http://localhost:4200/home because that is the redirect uri i have enabled in my google api dev console but i noticed upon returning from the login the app was trying to redirect (perhaps re-redirect) to http://localhost:4200/scope ('scope' is not one of my enabled redirect paths in my dev console).  Long story short i tried to work around this by defining the 'scope' path to just be a redirectTo: '/home' path in my app-routing module.  The app now lands back at home but then has other issues involving not being "logged in", i.e. isUserSignedIn() in UserService returns false and the GmailLabelService I have created cannot fetch labels because the user is not logged in, or at the very least the credentials are not stored;

have i not configured the ux_mode: "redirect" and redirect_uri: "http://localhost:4200/home"  configurations correctly in my app or is this a bug in which using ng-gapi to log in with redirect doesn't work in all cases?

I have attached my app module here for reference and will attach any other code requested.  I have REMOVED my clientId: as I'm not sure if this should be so public, but it is correctly in there.

you would think google would make it easier to use their api's from within Angular, but maybe this is the security via obscurity portion of their security operation...
...many many thanks to those who have helped to write this wrapper as i realize it is very non-trivial.

-Thank you for your time,
*starred*
<code>
...
import { UserService } from '@app/core/services/user.service';
import { GmailLabelService } from '@app/core/services/gmail-label.service';

import { HttpClientModule } from '@angular/common/http';

import {
    GoogleApiModule,
    GoogleApiService,
    GoogleAuthService,
    NgGapiClientConfig,
    NG_GAPI_CONFIG,
    GoogleApiConfig
} from "ng-gapi";

let gapiClientConfig: NgGapiClientConfig = {
    client_id: "REMOVED",
    discoveryDocs: ["https://www.googleapis.com/discovery/v1/apis/gmail/v1/rest"],
    scope: 'https://www.googleapis.com/auth/gmail.readonly',
    ux_mode: "redirect",
    redirect_uri: "http://localhost:4200/home"
};

@NgModule({
  declarations: [
    AppComponent,
  ],
  imports: [
    BrowserModule,
    BrowserAnimationsModule,

    //  CoreModule,
    SharedModule,
    AuthModule,
    AdminPanelModule,
    HttpClientModule,

    GoogleApiModule.forRoot({
      provide: NG_GAPI_CONFIG,
      useValue: gapiClientConfig
    }),

    AppRoutingModule, // this must always come last
  ],
  providers: [UserService, GmailLabelService],
  bootstrap: [AppComponent]
})
export class AppModule {
  constructor() {
  }
}
...
</code>

**Comment by rubenCodeforges** [comment]
Hi,  thanks for trying the lib.

Have you configured the localhost url inside your admin console ( google )  ?

**Comment by SanFranciscoSunrise** [comment]
Yes. I have both http://localhost:4200 and http://localhost:4200/home configured in the Google dev console. Is the redirect_uri a feature that you have seen work before?  I'm thinking I'm missing a step, maybe something about clearing the hash in the browser navigation bar after login?

**Comment by Ruben-E** [comment]
@SanFranciscoSunrise Looks like i'm running into the same issue. Have you already found a solution since you closed the issue?

**Comment by SanFranciscoSunrise** [comment]
I do not remember closing the issue. It should still be open. Do issues auto close after sometime? I have yet to have time experimenting with clearing the hash on the navigation bar with window.location.hash = ' ' as seen in this tutorial https://auth0.com/blog/amp/ngrx-authentication-tutorial/  I would definitely love for this redirect to work for ng-gapi as I am using Google's gmail api with ng-gapi help. Did you also run into it trying to redirect to scope? Please post if you solve.

**Comment by SanFranciscoSunrise** [comment]
OK, one step closer.  by setting { useHash: false } in the RouterModule.forRoot() method
```
@NgModule({
  imports: [RouterModule.forRoot(
    routes,
    { useHash: false }
  )],
  exports: [RouterModule]
})
export class AppRoutingModule { }
```
we at least don't get the error about 'scope' however now the service requires a different sign-in flow as it breaks away from the .then in

```
public signIn() {
        this.googleAuthService.getAuth().subscribe((auth) => {
            auth.signIn().then(res => this.signInSuccessHandler(res), 
                                             err => this.signInErrorHandler(err));
        });
    }
```

and does return us to out http://localhost:4200/callback component with the client_id, scopes, and some other data in the url hash fragment....

now how do we take that client_id from the hash fragment and get the access_token from the redirected page???

**Comment by Ruben-E** [comment]
Well apparently [Ghostery](https://www.ghostery.com/) was doing some magic stuff. The redirect suddenly works fine with Ghostery disabled. It still doesn't work on my phone though. That's why I needed a redirect in the first place. To support phones.

My source code lives here: https://github.com/Ruben-E/kilometeradministratie/tree/master/src/app please compare it with yours to proceed one step further. I'm going to investigate my the redirect on the phone isn't working.

**Comment by rubenCodeforges** [comment]
@SanFranciscoSunrise @Ruben-E  Hey guys , sorry for delayed response , i was a bit busy recently.

Ive modified the stackblitz example for your case

https://ng-gapi-example.stackblitz.io - see the dev console
and the version with code https://stackblitz.com/edit/ng-gapi-example?file=app%2Fapp.module.ts , please note that the version with code will throw a error when you try to open the google auth frame as it is launching from a frame a new frame , thats why you should open the full version https://ng-gapi-example.stackblitz.io, but the code should give you insights on what is going on.

Basically , the route.fragment.subcribe is where you should look for you fragments , how you parse them its up to you , also note that app.module i did not disabled the hash strategy and it works perfectly.

Things that might be block , if you have a scope on your app that need authorization by google , it want work until authorized. See your google console for warnings. In my case i used a public scope for profile info.
I hope that helps.
